tianshou/data/utils/segtree.py

```diff
@@ -0,0 +1,117 @@
+import numpy as np
+from typing import Union, Optional
+# from numba import njit
+
+
+class SegmentTree:
+    """Implementation of Segment Tree. The procedure is as follows:
+
+    1. Find out the smallest n which safisfies ``size <= 2^n``, and let \
+    ``bound = 2^n``. This is to ensure that all leaf nodes are in the same \
+    depth inside the segment tree.
+    2. Store the original value to leaf nodes in ``[bound:bound * 2]``, and \
+    the union of elementary to internal nodes in ``[1:bound]``. The internal \
+    node follows the rule: \
+    ``value[i] = operation(value[i * 2], value[i * 2 + 1])``.
+    3. Update a node takes O(log(bound)) time complexity.
+    4. Query an interval [l, r] with the default operation takes O(log(bound))
+
+    :param int size: the size of segment tree.
+    :param str operation: the operation of segment tree. Choose one of "sum",
+        "min" and "max", defaults to "sum".
+    """
+
+    def __init__(self, size: int,
+                 operation: str = "sum") -> None:
+        bound = 1
+        while bound < size:
+            bound <<= 1
+        self._bound = bound
+        assert operation in ["sum", "min", "max"], \
+            f"Unknown operation {operation}."
+        (self._op, self._init_value) = {
+            "sum": (np.sum, 0.),
+            "min": (np.min, np.inf),
+            "max": (np.max, -np.inf),
+        }[operation]
+        self._value = np.zeros([bound << 1]) + self._init_value
+
+    def __len__(self):
+        return self._bound
```

Comment:
Why is it ``self._bound`` rather than ``size``? I think ``self._bound`` is something internal. For users, they just know they have ``size`` elements.